topostats/tracing/skeletonize.py

```diff
@@ -2,78 +2,641 @@
+
+    def _binaryThinCheck_dsharp(self):
+        """Check if p2, p6 or p8 is 0 - seems very specific
+
+        Returns
+        -------
+        bool
+            if p2, p6 or p8 is 0.
+        """
+        return self.p2 * self.p6 * self.p8 == 0
+
+    def finalSkeletonisationIteration(self):
+        """A final skeletonisation iteration that removes "hanging" pixels.
+        Examples of such pixels are:
+                    [0, 0, 0]               [0, 1, 0]            [0, 0, 0]
+                    [0, 1, 1]               [0, 1, 1]            [0, 1, 1]
+            case 1: [0, 1, 0]   or  case 2: [0, 1, 0] or case 3: [1, 1, 0]
+
+        This is useful for the future functions that rely on local pixel environment
+        to make assessments about the overall shape/structure of traces"""
+
+        remaining_coordinates = np.argwhere(self.mask).tolist()
+
+        for x, y in remaining_coordinates:
+            self.p7, self.p8, self.p9, self.p6, self.p2, self.p5, self.p4, self.p3 = self.getLocalPixelsBinary(
+                self.mask, x, y
+            )
+
+            # Checks for case 1 pixels
+            if self._binaryThinCheck_b_returncount() == 2 and self._binaryFinalThinCheck_a():
+                self.mask[x, y] = 0
+            # Checks for case 2 pixels
+            elif self._binaryThinCheck_b_returncount() == 3 and self._binaryFinalThinCheck_b():
+                self.mask[x, y] = 0
+
+    def _binaryFinalThinCheck_a(self):
+        """Assess if local area has 4-connectivity.
+
+        Returns
+        -------
+        bool
+            T/F if any neighbours of the 4-connections have a near pixel.
+        """
+        return 1 in (self.p2 * self.p4, self.p4 * self.p6, self.p6 * self.p8, self.p8 * self.p2)
+
+    def _binaryFinalThinCheck_b(self):
+        """Assess if local area 4-connectivity is connected to multiple branches.
+
+        Returns
+        -------
+        bool
+            T/F if any neighbours of the 4-connections have a near pixel.
+        """
+        return 1 in (
+            self.p2 * self.p4 * self.p6,
+            self.p4 * self.p6 * self.p8,
+            self.p6 * self.p8 * self.p2,
+            self.p8 * self.p2 * self.p4,
+        )
+
+    @staticmethod
+    def getLocalPixelsBinary(binary_map, x, y) -> np.ndarray:
+        """Get the values of the pixels in the local 8-connectivit area around
+        the coordinate described by x and y.
+
+        Parameters
+        ----------
+        binary_map: np.ndarray
+            The binary array containing the grains.
+        x: int
+            An x coordinate within the binary map.
+        y: int
+            A y coordinate within the binary map.
+
+        Returns
+        -------
+        np.ndarray
+            A flattened 8-long array describing the values in the binary map
+            around the x,y point
+        """
+        local_pixels = binary_map[x - 1 : x + 2, y - 1 : y + 2].flatten()
+        # [[p7, p8, p9],    [[0,1,2],
```

Comment:
These lines can be removed if no longer needed.

Reply:
It's a description to show what each of the positions: p2, p3... etc mean. I've moved this into the docstring as I think it helps with understanding